### PR TITLE
[SecurityBundle] Fix configuring OIDC user info token handler client

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/18354

Cf. https://github.com/symfony/symfony/pull/50453#discussion_r1210028968

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php [hunk 1]
```diff
@@ -61,14 +45,9 @@ public function addConfiguration(NodeBuilder $node): void
                         ->info('Claim which contains the user identifier (e.g.: sub, email..).')
                         ->defaultValue('sub')
                     ->end()
-                    ->arrayNode('client')
+                    ->scalarNode('client')
                         ->info('HttpClient to call the OIDC server.')
                         ->isRequired()
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: why do we keep this? can't we default to `http_client`?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php [hunk 2]
```diff
@@ -27,23 +26,8 @@ class OidcUserInfoTokenHandlerFactory implements TokenHandlerFactoryInterface
...
         $tokenHandlerDefinition = $container->setDefinition($id, new ChildDefinition('security.access_token_handler.oidc_user_info'));
+        $tokenHandlerDefinition->replaceArgument(0, new Reference($config['client']));
         $tokenHandlerDefinition->replaceArgument(2, $config['claim']);
-
-        // Create the client service
-        if (!isset($config['client']['id'])) {
-            $clientDefinitionId = 'http_client.security.access_token_handler.oidc_user_info';
-            if (!ContainerBuilder::willBeAvailable('symfony/http-client', HttpClient::class, ['symfony/security-bundle'])) {
-                $container->register($clientDefinitionId, 'stdClass')
-                    ->addError('You cannot use the "oidc_user_info" token handler since the HttpClient component is not installed. Try running "composer require symfony/http-client".');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php:37]: we might still need to register this special def when the client is not installed, isn't it?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php (hunk 1 above)
Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: Cause we need the base_uri of the OIDC server:
```yaml
# config/packages/framework.yaml
framework:
    http_client:
        scoped_clients:
            oidc.client:
                base_uri: 'https://www.example.com/realms/demo/protocol/openid-connect/userinfo'

# config/packages/security.yaml
security:
    firewalls:
        main:
            access_token:
                token_handler:
                    oidc_user_info:
                        client: oidc.client
```

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php (hunk 2 above)
Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php:37]: `client` option receives an http_client service id, so I guess the user installed symfony/http-client to declare it. Do we need to double-check it here?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: are there any other options that might be useful to configure?
because if not, we might just need `oidc_user_info: <base-uri-as-string>`

To do so, I'd suggest defining an abstract service, let's name it `security.access_token_handler.oidc_user_info.http_client`, defined as such:

```php
->set('security.access_token_handler.oidc_user_info.http_client', HttpClientInterface::class)
	->abstract()
	->factory([service('http_client'), 'withOptions'])
```

and in the OidcUserInfoTokenHandlerFactory, we'd create a ChildDefinition of this service and set it's argument to `['base_uri' => $config['base_uri']]`.

Does this make sense to you?

Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: `oidc_user_info` can also accept a `claim` optional parameter:
```yaml
oidc_user_info:
  claim: email # default: sub
  client: oidc.client # required
```

I can allow `oidc_user_info: <base-uri>` and normalize it, but I find this syntax more confusing than setting the `oidc_user_info.client` parameter. But this is just a question of point of view. WDYT?

Anyway, if a user wants to change the `claim`, we still need a parameter to set the base_uri (or a custom client).

Comment (nicolas-grekas) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: I think the DX will be much better with the changes I suggested.
We can accept both:
```yaml
oidc_user_info: 'https://...'
# same as:
oidc_user_info:
  claim: sub
  base_uri: 'https://...'
```

Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: We should keep the `client` option in case the dev wants to manually configure its client externally (e.g.: `framework.http_client`, or any service manually created).

In your approach, the `base_uri` option would be used to dynamically create the client (and set the `client` option), and the `client` option would receive a HttpClient service id:
```yaml
# HttpClient dynamically created with "base_uri" only
oidc_user_info: https://...

# HttpClient dynamically created with "base_uri" only
oidc_user_info:
  claim: sub
  base_uri: https://... # "base_uri" is not compatible with "client" option

# HttpClient manually created with custom options
oidc_user_info:
  claim: sub
  client: oidc.client
```

But I think it's making this configuration and DX more complex... IMHO, I would prefer to set the HttpClient options in the `client` option like it was earlier (but this time, using the abstract definition you suggested):
```yaml
# HttpClient manually created with custom options
oidc_user_info:
  claim: sub
  client:
    base_uri: https://...
    # other HttpClient options

# HttpClient dynamically created with "base_uri" only
oidc_user_info: https://...
```

Comment (nicolas-grekas) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: I'm fine adding another option to set the service id if that's something that'd be useful, but I don't think allowing other http_client options is desired (as explained in https://github.com/symfony/symfony/pull/50453#discussion_r1210028968)
But being able to set only the base_uri looks the best DX to me. I can do the code if you're not comfortable with it of course. Let me know!

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php (hunk 2 above)
Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php:37]: Solved with the addition of `base_uri` option

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php (hunk 1 above)
Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: Fixed in my last commit

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Security/Factory/AccessTokenFactoryTest.php [hunk 3]
```diff
@@ -103,7 +120,19 @@ public function testOidcUserInfoTokenHandlerConfigurationWithClientCreation()
...
+
+        $expected = [
+            'index_0' => (new ChildDefinition('security.access_token_handler.oidc_user_info.http_client'))
+                ->replaceArgument(0, ['base_uri' => 'https://www.example.com/realms/demo/protocol/openid-connect/userinfo']),
+            'index_2' => 'sub',
+        ];
+        $this->assertEquals($expected, $container->getDefinition('security.access_token_handler.firewall1')->getArguments());
+    }
+
+    public function getOidcUserInfoConfiguration(): iterable
```
Comment (mtarld) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Security/Factory/AccessTokenFactoryTest.php]: Must be static in order to be compatible with the next PHPUnit version

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php [hunk 4]
```diff
@@ -56,19 +52,21 @@ public function addConfiguration(NodeBuilder $node): void
         $node
             ->arrayNode($this->getKey())
                 ->fixXmlConfig($this->getKey())
+                ->beforeNormalization()
+                    ->ifString()
+                    ->then(static fn ($v) => ['claim' => 'sub', 'base_uri' => $v])
+                ->end()
                 ->children()
+                    ->scalarNode('base_uri')
+                        ->info('Base URI of the OIDC server.')
```
Comment (vincentchalamon) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]:
```suggestion
                        ->info('Base URI of the userinfo endpoint on the OIDC server.')
```

**Review by nicolas-grekas — APPROVED** [review]
I tweaked the implementation a bit, see
https://github.com/symfony/symfony/compare/8e7452011403b4bbe67a4d2abd12dde3df95f0c3..8b7fdc108237bd8e88c59aa0282aa6d36d9ae7ad

TL;DR, the `base_uri` option is required, and the `client` option can be used optionally to configure any other options.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php [hunk 5]
```diff
@@ -56,19 +52,21 @@ public function addConfiguration(NodeBuilder $node): void
...
+                        ->isRequired()
+                    ->end()
                     ->scalarNode('claim')
-                        ->info('Claim which contains the user identifier (e.g.: sub, email..).')
+                        ->info('Claim which contains the user identifier (e.g. sub, email, etc.).')
                         ->defaultValue('sub')
```
Comment (mtarld) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php:67]: You can add `isRequired` and `canntBeEmpty` as well IMHO

Comment (nicolas-grekas) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php:67]: cannotBeEmpty why not, can you take care of @mtarld's comments and squash @vincentchalamon please?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php [hunk 6]
```diff
@@ -60,10 +60,13 @@ public function addConfiguration(NodeBuilder $node): void
                     ->scalarNode('base_uri')
                         ->info('Base URI of the userinfo endpoint on the OIDC server.')
                         ->isRequired()
+                        ->cannotBeEmpty()
                     ->end()
                     ->scalarNode('claim')
                         ->info('Claim which contains the user identifier (e.g. sub, email, etc.).')
                         ->defaultValue('sub')
+                        ->isRequired()
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: does this one needs to be required if it has a default value?

Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcUserInfoTokenHandlerFactory.php]: Nop, removed in a squashed commit. Tests revealed the bug too.
Thanks for noticing :wink:

**Comment by nicolas-grekas** [comment]
Thank you @vincentchalamon.
